pipit/readers/nsight_reader.py

```diff
@@ -0,0 +1,129 @@
+        df["Time"] = df["Start (ns)"]
+        df["Graph_Node"] = None
+        df["Level"] = None
+
+        for i in range(len(df)):
+            node = graph.get_node(df.iloc[i]["RangeStack"])
+            df.at[i, "Graph_Node"] = node
+            df.at[i, "Level"] = node.get_level()
+
+        # Create new columns for df2 with end time
+        df2["Event Type"] = "Exit"
+        df2["Time"] = df2["End (ns)"]
+
+        for i in range(len(df2)):
+            node = graph.get_node(df2.iloc[i]["RangeStack"])
+            df2.at[i, "Graph_Node"] = node
+            df2.at[i, "Level"] = node.get_level()
+
+        df2.Level = df2.Level.astype(int)
+
+        # Combine dataframes together
+        df = pd.concat([df, df2])
+
+        # Tidy Dataframe
+        df.drop(["Start (ns)", "End (ns)"], axis=1, inplace=True)
+
+        df.sort_values(by="Time", ascending=True, inplace=True)
+
+        df.reset_index(drop=True, inplace=True)
+
+        return pipit.trace.Trace(None, df)
+
+    """
+    This call graph function takes in a dataframe and shows the user
+    how the call graph would look like within their program.
+    """
+
+    def call_graph(self):
+        df = self.df
+
+        # Creating call graph
+        call_graph = graph.Graph()
+
+        # Creating a stack
+        stack = []
+
+        for i in range(len(df)):
+            # create a root
+            if call_graph.is_empty():
+                root = graph.Node(-1, df.iloc[i]["Name"], None)
+                root.add_calling_contex_id(df.iloc[i]["RangeStack"])
+                root.add_time(df.iloc[i]["Start (ns)"], df.iloc[i]["End (ns)"])
+
+                # add root to tree
+                call_graph.add_root(root)
+                call_graph.add_to_map(root.calling_context_ids[0], root)
+
+                # add root stack
+                stack.append(df.iloc[i])
+
+            # If existing root
+            else:
+                curr_start = df.iloc[i]["Start (ns)"]
+                prev_end = stack[len(stack) - 1]["End (ns)"]
+
+                # CASE 1 Function in another function
+                if prev_end > curr_start:
+                    parent = call_graph.get_node(stack[len(stack) - 1]["RangeStack"])
+                    node = graph.Node(-1, df.iloc[i]["Name"], parent)
+                    node.add_calling_contex_id(df.iloc[i]["RangeStack"])
+                    node.add_time(df.iloc[i]["Start (ns)"], df.iloc[i]["End (ns)"])
+                    parent.add_child(node)
+
+                    call_graph.add_to_map(node.calling_context_ids[0], node)
+
+                    # Adding child function to stack
+                    stack.append(df.iloc[i])
+
+                # Case #2 Function is outside of the previous function
+                else:
+                    while stack[len(stack) - 1]["End (ns)"] < curr_start:
+                        stack.pop()
+
+                    parent = call_graph.get_node(stack[len(stack) - 1]["RangeStack"])
```

Comment:
If the DataFrame consisted of a root followed by another root, I think there could be an index error here.

Ex:
root1 happens from 1 to 10 and root2 happens from 11 to 20

In this case, the first root would be removed from the stack and you would be trying to index the first element of an empty stack.

Reply:
I agree. Will look into it.